Runtime clock scaling feature support in XRT

rajkumar-xilinx opened the pull request:
Hi,

This change contains following changes:
-Added new API named xclSetThresholdValues() in xocl/api/xlnx XRT core to configure the power and temperature threshold/target values.
-Added new ioctl in mgmtpf driver for this configuration settings.
-mgmtpf driver will redirect the function to XMC suddev driver, where actual CMC register programming takes place
-No impact on the current platforms since this feature will uses featureROM bitmap data to enable this new feature.

More information about this feature is available @ http://confluence.xilinx.com/display/~spyla/Clock+Frequency+Scaling

gbuildx commented:
Can one of the admins verify this patch?

maxzhen commented:
ok to test

gbuildx commented:
Build Passed!

rajkumar-xilinx commented on src/runtime_src/driver/xclng/drm/xocl/subdev/xmc.c:
```diff
@@ -168,11 +184,44 @@ struct xocl_xmc {
+  ctrl &= ~XMC_CLOCK_SCALING_TEMP_REG_MASK;
+  ctrl |= (obj->temperature & XMC_CLOCK_SCALING_TEMP_REG_MASK);
+  WRITE_CLK(xmc, ctrl, XMC_CLOCK_SCALING_TEMP_REG);
+}
```
Yes, it's an open action item. I forgot to add TODO statement as a comment at this place in the code. I updated TODO comment here now. The power/temperature spec limits has to be read from featureROM field (which is being tracked in CR-1029672. Once the featureROM changes are ready, I will add the check to prevent setting up power/temp values more than spec limits.

rajkumar-xilinx commented:
> 1. I'd expect some check in driver to prevent user from setting threshold to something outside of spec, but I don't see where this is done?
> 2. We allow user to set threshold, but we don't have API for user to read them back? Or maybe we don't think they ever need to? Do we want to show threshold in xbutil query?
> 3. What are default thresholds? How do we restore the default value after messed up by some user?

Point#1 will be taken care once CR-1029672 is fixed. I added a TODO comment at power/temp configuration code.
Point#2: Planning to show the power and temperature board limits through sysfs. This task will also taken care along with Point#1. In addition, the current power/temperature values can be shown through xbutil query or sysfs entry. I will add this change.
Point#3: Restoring the values to default: I will explore on this task. Once the kernel execution is finished (or host program finished) then xrt should restore the values to default (i.e. max power/temp spec limits). I will go through XRT and will find out how to add this change.

gbuildx commented:
Build Passed!

rajkumar-xilinx commented:
> Finally a very sought after feature! My high level feedback:
> 
> 1. Threshold changes should be done through sysfs; ioctls are not necessary for this purpose
> 2. Only root should be allowed to change the values; driver should validate the request
> 3. We do not need a formal XRT API for this functionality
> 4. Expose the functionality through xbmgmt

Hi Sonal,
 Thanks for the feedback, This can be implemented either from sysfs or xbmgmt or both. Below are my thoughts about these approaches.

1) Someone can update threshold values without loading an xclbin, Any other user who uses that device is going to be effected.  
2) If User1 started an application and User2 can change the values thru sysfs/xbmgmt. 
3) If User can change the threshold values only from Application, then no user can update it as device is locked. XRT can restore the default values once application is finished

Please let me know if i am missing anything

sonals commented:
@rajkumar-xilinx please look at Linux standard way of handling frequency scaling through sysfs:

xsjsonals5570:~>tree /sys/devices/system/cpu/cpu0/cpufreq
/sys/devices/system/cpu/cpu0/cpufreq
├── affected_cpus
├── cpuinfo_max_freq
├── cpuinfo_min_freq
├── cpuinfo_transition_latency
├── energy_performance_available_preferences
├── energy_performance_preference
├── related_cpus
├── scaling_available_governors
├── scaling_cur_freq
├── scaling_driver
├── scaling_governor
├── scaling_max_freq
├── scaling_min_freq
└── scaling_setspeed

0 directories, 14 files
xsjsonals5570:~>

The clock scaling framework for Alveo should mimic this architecture. 
There is no single process model starting from 2019.1 -- XRT defaults to multi-process behavior.

gbuildx commented:
Build Passed!

gbuildx commented:
Build Passed!

rajkumar-xilinx commented:
Hi, I have updated the patch for configuring power/temp threshold values through sysfs only as per sugesstions. I have added clock scaling entries under xmc sysfs node, which is already available under pcie device node.

#ls /sys/devices/pci0000:3a/0000:3a:00.0/0000:3b:00.0/xmc.m.15104/scaling_*
scaling_cur_power  scaling_cur_temp  scaling_enabled

gbuildx commented:
Build Passed!

maxzhen commented:
> 
> 
> Hi, I have updated the patch for configuring power/temp threshold values through sysfs only as per sugesstions. I have added clock scaling entries under xmc sysfs node, which is already available under pcie device node.
> 
> #ls /sys/devices/pci0000:3a/0000:3a:00.0/0000:3b:00.0/xmc.m.15104/scaling_*
> scaling_cur_power
> scaling_cur_temp
> scaling_enabled
> scaling_governor
> scaling_target_power
> scaling_target_temp

Why do you need scaling_cur_temp/scaling_cur_power? I thought they are already available and consumed by xbutil query, no?

rajkumar-xilinx commented:
> > Hi, I have updated the patch for configuring power/temp threshold values through sysfs only as per sugesstions. I have added clock scaling entries under xmc sysfs node, which is already available under pcie device node.
> > #ls /sys/devices/pci0000:3a/0000:3a:00.0/0000:3b:00.0/xmc.m.15104/scaling_*
> > scaling_cur_power
> > scaling_cur_temp
> > scaling_enabled
> > scaling_governor
> > scaling_target_power
> > scaling_target_temp
> 
> Why do you need scaling_cur_temp/scaling_cur_power? I thought they are already available and consumed by xbutil query, no?

Hi Max,
Yes, these values can be ready by xbutil query command also. I thought keeping all scaling parameters in one place will be convenient to user. I will remove these settings if they are redundant.
Also, can you please let me know if the other changes looks ok.

maxzhen commented:
> 
> 
> > > Hi, I have updated the patch for configuring power/temp threshold values through sysfs only as per sugesstions. I have added clock scaling entries under xmc sysfs node, which is already available under pcie device node.
> > > #ls /sys/devices/pci0000:3a/0000:3a:00.0/0000:3b:00.0/xmc.m.15104/scaling_*
> > > scaling_cur_power
> > > scaling_cur_temp
> > > scaling_enabled
> > > scaling_governor
> > > scaling_target_power
> > > scaling_target_temp
> > 
> > 
> > Why do you need scaling_cur_temp/scaling_cur_power? I thought they are already available and consumed by xbutil query, no?
> 
> Hi Max,
> Yes, these values can be ready by xbutil query command also. I thought keeping all scaling parameters in one place will be convenient to user. I will remove these settings if they are redundant.
> Also, can you please let me know if the other changes looks ok.

Some of your changes in driver still use spaces for indent. Please fix them. I have no other comments. For things you don't plan to fix right away, please file CRs and assign them to yourself so that we don't forget.

gbuildx commented:
Build Failed! :(

rajkumar-xilinx commented:
> Build Failed! :(

Hi,
can you please re-trigger the build test. One of the scripts failed with below error message. 
>>>
22:53:16 [xrt-ubuntu16.04] Running shell script
22:53:17 + /tools/xgs/bin/sudo apt-get remove -y xrt xrt-aws
22:53:17 sudo: start_tls specified but LDAP libs do not support ldap_start_tls_s() or ldap_start_tls_s_np()
22:53:17 E: Could not get lock /var/lib/dpkg/lock-frontend - open (11: Resource temporarily unavailable)
22:53:17 E: Unable to acquire the dpkg frontend lock (/var/lib/dpkg/lock-frontend), is another process using it?
>>>

rajkumar-xilinx commented:
retest this please

maxzhen commented:
retest this please

gbuildx commented:
Build Passed!

gbuildx commented:
Build Passed!

rajkumar-xilinx commented:
Below are list of open Tasks for feature to be fully functional.
http://jira.xilinx.com/browse/SPT-133
http://jira.xilinx.com/browse/SPT-134
http://jira.xilinx.com/browse/CR-1029672
http://jira.xilinx.com/browse/SPT-87

maxzhen commented:
Merging now since the basic functionality is ready. Please make sure you address the pending issues as listed above in a timely manner.
